src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,329 @@
+        {
+            _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
+            _map = new ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
+            _eventQueueTask = new Task(OnStart, options);
+            _eventQueueTask.Start();
+            _ = RemoveExpiredValuesPeriodically(TimeSpan.FromMinutes(5));
+        }
+
+        private void OnStart()
+        {
+            while (!_disposed)
+            {
+                try
+                {
+                    if (_eventQueue.TryTake(out var action, 50))
+                        action.Invoke();
+                }
+                catch (Exception ex)
+                {
+                    LogHelper.LogWarning(LogHelper.FormatInvariant(LogMessages.IDX10900, ex));
+                }
+               
+            }
+        }
+
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// FOR TESTING PURPOSES ONLY.
+        /// </summary>
+        internal void WaitForProcessing()
+        {
+            while (_eventQueue.Count != 0)
+                continue;
+
+            return;
+        }
+
+        internal int RemoveExpiredValues()
+        {
+            int numItemsRemoved = 0;
+            var node = _doubleLinkedList.First;
+            while (node != null)
+            {
+                var nextNode = node.Next;
+                if (node.Value.ExpirationTime < DateTime.UtcNow)
+                {
+                    _doubleLinkedList.Remove(node);
+                    _map.TryRemove(node.Value.Key, out _);
+                    numItemsRemoved++;
+                }
+
+                node = nextNode;
+            }
+
+            return numItemsRemoved;
+        }
+
+        internal void RemoveLRUs()
+        {
+            // use the _capacity for the newCacheSize calculation in the case where the cache is experiencing overflow
+            int currentCount = _map.Count <= _capacity ? _capacity : _map.Count;
+            var newCacheSize = currentCount - (int)(currentCount * _compactionPercentage);
+            while (_map.Count > newCacheSize && _doubleLinkedList.Count > 0)
+            {
+                var lru = _doubleLinkedList.Last;
+                _map.TryRemove(lru.Value.Key, out _);
+                _doubleLinkedList.Remove(lru);
+            }
+        }
+
+        async Task RemoveExpiredValuesPeriodically(TimeSpan interval)
+        {
+            while (!_disposed)
+            {
+                _eventQueue.Add(() => RemoveExpiredValues());
+                await Task.Delay(interval).ConfigureAwait(false);
+            }
+        }
```

Comment:
I'm wondering if this is causing a dispose of resources challenge if there are several LRUCaches created as the thread is waiting for 5 minutes. It can be considered to add the capability to the OnStart process (and do a simple comparison there on time instead of this). However if we haven't seen any issues in tests so far there is not likely a need to change this now.

Reply:
ack. I think we should have a task to move this code to onstart

---
In reply to: [590681156](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1605#discussion_r590681156) [](ancestors = 590681156)

Reply:
#1609